lib/screens/v2/disruption_diagram/builder.ex

```diff
@@ -0,0 +1,967 @@
+          {full, abbrev} = Map.fetch!(stop_id_to_name, stop_id)
+
+          %StopSlot{
+            id: stop_id,
+            label: %{full: full, abbrev: abbrev},
+            home_stop?: stop_id == localized_alert.location_context.home_stop,
+            disrupted?: stop_id in informed_stop_ids,
+            terminal?: false
+          }
+        end)
+        |> adjust_ends(line, branch)
+
+      init_metadata = %Metadata{
+        line: line,
+        branch: branch,
+        effect: localized_alert.alert.effect,
+        # These will get the correct values during the first `recalculate_metadata` run below.
+        home_stop: -1,
+        first_disrupted_stop: -1,
+        last_disrupted_stop: -1
+      }
+
+      builder =
+        %__MODULE__{sequence: slot_sequence, metadata: init_metadata}
+        |> recalculate_metadata()
+        |> split_end_stops()
+
+      {:ok, builder}
+    end
+  end
+
+  @doc """
+  Reverses the builder's internal stop sequence, so that the last stop comes first and vice versa.
+
+  This is helpful for cases where the disruption diagram lists stops in the opposite order of
+  the direction_id=0 route order, e.g. in Blue Line diagrams where we show Bowdoin first but
+  direction_id=0 has Wonderland listed first.
+  """
+  @spec reverse(t()) :: t()
+  def reverse(%__MODULE__{} = builder) do
+    %{
+      builder
+      | sequence: Vector.reverse(builder.sequence),
+        # The ends swap places, and also have their elements flipped.
+        left_end: Vector.reverse(builder.right_end),
+        right_end: Vector.reverse(builder.left_end)
+    }
+    |> recalculate_metadata()
+  end
+
+  @doc """
+  Tries to omit stops from the given region, replacing them with a labeled "blank" slot, or two in rare cases.
+  `target_slots` gives the desired number of remaining slots in the region after omission.
+
+  Stops are omitted from the center of the region, unless that would result
+  in the omission of the home stop or a bypassed stop.
+  In that case, we try to find another segment, or segments, of stops to omit, staying as close to the center as possible.
+
+  Returns an error result if it's not possible to omit the required number of stops without
+  also omitting the home stop or a bypassed stop.
+  """
+  @spec try_omit_stops(t(), :closure | :gap, pos_integer()) ::
+          {:ok, t()} | {:error, reason :: String.t()}
+  def try_omit_stops(builder, region, target_slots)
+
+  def try_omit_stops(%__MODULE__{} = builder, :closure, target_closure_slots) do
+    try_omit(builder, closure_indices(builder), target_closure_slots)
+  end
+
+  def try_omit_stops(%__MODULE__{} = builder, :gap, target_gap_stops) do
+    try_omit(builder, gap_indices(builder), target_gap_stops)
+  end
+
+  # Returns a sorted vector containing indices of stops that can't be omitted from the closure region.
+  defp get_important_indices(builder) do
+    closure_first..closure_last//1 = closure = closure_indices(builder)
```

Comment:
How do I interpret the first part of this line?

Reply:
`closure = closure_indices(builder)` - binds the whole range to `closure`
`closure_first..closure_last//1 = closure` - binds the first and last values of the range to `closure_first` and `closure_last`, and asserts that the step of the range is 1

This is equivalent to pattern matches in `case` and such like:
```ex
case %{foo: :bar} do
  %{foo: foo_variable_binding} = map_containing_foo -> ...
end
```
which bind multiple variables to different parts/nesting levels of a compound value.